CSharp/Library/QnAMaker/QnAMakerDialog.cs

```diff
@@ -1,110 +1,225 @@
+        {
+            var message = await argument;
+            var qnaMakerResults = default(QnAMakerResults);
+
+            if (message != null && !string.IsNullOrEmpty(message.Text))
+            {
+                var tasks = this.services.Select(s => s.QueryServiceAsync(message.Text)).ToArray();
+
+                if (tasks.Any())
+                {
+                    var maxValue = tasks.Max(x => x.Result.Answers[0].Score);
+                    qnaMakerResults = tasks.First(x => x.Result.Answers[0].Score == maxValue).Result;
+
+                    var filteredQnAResults = new List<QnAMakerResult>();
+                    foreach (var qnaMakerResult in qnaMakerResults.Answers)
+                    {
+                        qnaMakerResult.Score /= 100;
+                        if (qnaMakerResult.Score >= qnaMakerResults.ServiceCfg.ScoreThreshold)
+                        {
+                            qnaMakerResult.Answer = HttpUtility.HtmlDecode(qnaMakerResult.Answer);
+                            filteredQnAResults.Add((qnaMakerResult));
+                        }
+                    }
+
+                    qnaMakerResults.Answers = filteredQnAResults;
+
+                    if (qnaMakerResults != null && qnaMakerResults.Answers != null && qnaMakerResults.Answers.Count > 0)
+                    {
+                        if (this.IsConfidentAnswer(qnaMakerResults))
+                        {
+                            await this.RespondFromQnAMakerResultAsync(context, message, qnaMakerResults);
+                            await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                        }
+                        else
+                        {
+                            var feedbackRecord = new FeedbackRecord { UserId = message.From.Id, UserQuestion = message.Text };
+                            context.PrivateConversationData.SetValue("qnaMakerResults", qnaMakerResults);
+                            context.PrivateConversationData.SetValue("feedbackRecord", feedbackRecord);
+                            await this.QnAFeedbackStepAsync(context, qnaMakerResults);
+                        }
+                    }
+                    else
+                    {
+                        await context.PostAsync(qnaMakerResults.ServiceCfg.DefaultMessage);
+                        await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                    }
+                }
+                else
+                {
+                    await context.PostAsync(qnaMakerResults.ServiceCfg.DefaultMessage);
+                    await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                }
+            }
+            else
+            {
+                await context.PostAsync(qnaMakerResults.ServiceCfg.DefaultMessage);
+                await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);

```

Comment:
it seems that all the 3 else sections above are doing the same thing. probably better to refactor the conditions to end up with one single else condition rather than 3.